Git Username and Password Binding Update (#4423)
jenkins-infra/jenkins.io: content/projects/gsoc/2021/projects/git-credentials-binding.adoc

They will be able to use the git command line options of their choice, including `--single-branch` and `--recurse-submodules`.
Users will be able to run authenticated git commands in their Jenkins Pipelines without modifying the git plugin.

=== Implementation
Both `gitUsernamePassword` and `gitSshPrivateKey` bindings use the https://plugins.jenkins.io/credentials/[Credential Plugin]

> to retrieve user's credential using the Credentials API, the https://plugins.jenkins.io/credentials-binding/[Credential Binnding Plugin]

Reviewer: Usually prefer sentence per line and shorter sentences.  Optional
```suggestion
to retrieve user's credential using the Credentials API.
The https://plugins.jenkins.io/credentials-binding/[Credential Binnding Plugin]
```

is used to bind Git specific environment variables with programs to perform authentication on behalf of the user, without their interaction with the command-line.

==== Git Username and Password Binding

> The `gitUsernamePassword` implementation use the Jenkins username and password

Reviewer: Minor grammar (optional)
```suggestion
The `gitUsernamePassword` implementation uses the Jenkins username and password
```

> values retrieved through Credential API, to access a remote repository  over *HTTP* protocol. This binding use the
> `GIT_ASKPASS` environment variable to provide credentials requested by `Git CLI`, program specified by this variable is invoked with a suitable prompt on the command-line, and the user’s credential are read from its standard output.

Reviewer: Minor phrasing, optional
```suggestion
values retrieved through Credential API, to access a remote repository  over *HTTP* protocol.
This binding uses the `GIT_ASKPASS` environment variable to provide credentials requested by command line git.
The program specified by this variable is invoked with a suitable prompt on the command-line, and the user’s credential are read from its standard output.
```

> image::https://gist.githubusercontent.com/arpoch/96c75cc32afd0e50476e273bd0269a86/raw/7875f70561e1417721ec7f637fa6f160a27822a5/UsernameandPassword.png[Git Username and Password]

Reviewer: I believe that the preference is to include images in the repository rather than reference them as a gist from outside the repository.  Could you place a copy of this image in the correct images directoy?
```suggestion
image::https://gist.githubusercontent.com/arpoch/96c75cc32afd0e50476e273bd0269a86/raw/7875f70561e1417721ec7f637fa6f160a27822a5/UsernameandPassword.png[Git Username and Password]
```

IMPORTANT: By default, two variable bindings `GIT_USERNAME` and `GIT_PASSWORD`,

> bind with user's credential i.e. username and password respectively. These variable binding's are available in both freestyle project and pipeline script.

Reviewer: Not a possessive, don't need the single quote
```suggestion
bind with user's credential i.e. username and password respectively. These variable bindings are available in both freestyle project and pipeline script.
```

> The `gitUsernamePassword` binding only provides authentication support for `Git CLI` implementaton and other Git implementations inside the `git client` plugin such as `JGit` and `JGit with Apache HTTP Client` are not supported. The variable bindings `GIT_USERNAME` and `GIT_PASSWORD` are accessible in `Git CLI`, JGit` and `JGit with Apache HTTP Client` Git implementations.

Reviewer: Prefer two sentences.  Optional.  I removed the last sentence because I didn't understand what you were trying to describe there.  Can you describe that in more detail so that I better understand it?
```suggestion
The `gitUsernamePassword` binding only provides authentication support for the `Git CLI` implementation.
Other Git implementations inside the `git client` plugin such as `JGit` and `JGit with Apache HTTP Client` are not supported.
```
